### DISCUSSION TITLE
Best way to meter query resource usage

### DISCUSSION THREAD

**Discussion by markandpathak** [ClickHouse/ClickHouse#50762] [Q&A] [answered]
We plan to use Clickhouse Cloud in our upcoming product. In our case number of users will be querying our database. For each query we would like to meter the amount of RAM used and execution time and charge users for their usage. I am wondering that apart from polling `query_logs` table for memory and execution time for each query id executed, are there any other better ways through which I can achieve the same? Can I get this metrics in the response to query it self in like profile or log data?

I know that trace logs have peakMemory consumption and I've tried to extract this information from trace logs by setting `send_logs_level = trace` but logs are unstructured and it's difficult to parse memory consumption from trace logs programmatically. May be query log can be provided in structured manner for `send_logs_level`? I'm using clickhouse-go client.

**Comment by kapustor** [comment, marked as answer]
Hi @markandpathak , based on our experience using `query_log` for getting per query memory information is a good choice. In addition to the `memory_usage` field, consider the `ProfileEvents` that can store some useful metrics.
